SwipeView

jsuarezruiz opened the pull request:
### Description of Change ###

[Swiping on an element](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/swipe) to [reveal or execute contextual commands](https://material.io/design/components/lists.html#behavior) is an increasingly common paradigm, especially when working with lists.  **SwipeView** provides support for that interaction.

Android
![swipeview-droid2](https://user-images.githubusercontent.com/6755973/67285916-b5b28680-f4d8-11e9-8289-57902d6d87df.gif)

iOS

![swipeview-ios2](https://user-images.githubusercontent.com/6755973/67285920-b6e3b380-f4d8-11e9-8917-aec194e66004.gif)

UWP
![swipeview-uwp](https://user-images.githubusercontent.com/6755973/65323497-d02bd400-dba9-11e9-9faa-8869330753f9.gif)

fixes #3171

### API Changes ###

## SwipeItem

Represents an individual command in a SwipeView.

```csharp
public class SwipeItem : ContextItem
{
	public static readonly BindableProperty BackgroundColorProperty;
	public Color BackgroundColor { get; set; }
}
```

## SwipeItems

Represents a collection of SwipeItem objects

```csharp
public class SwipeItems : IList<SwipeItem>
{
	public static SwipeMode Mode { get; set; }
}
```

### Properties

| API | Description |
| ------------- | ------------- |
| SwipeMode | Gets or sets the SwipeMode of the item. |

### Properties

| API | Description |
| ------------- | ------------- |
| BackgroundColor| Gets or sets the background color of the item. |

### SwipeMode

Defines the possible swipe interactions

```csharp
public enum SwipeMode 
{
	Reveal, // Display additional context items which may be selected
	Execute // Immediately execute the associated command
}
```

### SwipeBehaviorOnInvoked

Defines constants that specify how a SwipeView behaves after a command is invoked.

```csharp
public enum SwipeBehaviorOnInvoked 
{
	Auto, // In Reveal mode, the SwipeView closes after an item is invoked. In Execute mode, the SwipeView remains open.
	Close, // The SwipeView closes after an item is invoked.
        RemainOpen //The SwipeView remains open after an item is invoked.
}
```
## SwipeView

Represents a container that provides access to contextual commands through touch interactions.

```csharp
public class SwipeView : ContentView
{
	public static readonly BindableProperty LeftItems;
	public SwipeItems LeftItems { get; set; }

	public static readonly BindableProperty RightItems;
	public SwipeItems RightItems { get; set; }	

	public static readonly BindableProperty TopItems;
	public SwipeItems TopItems { get; set; }
	
	public static readonly BindableProperty BottomItems;	
	public SwipeItems BottomItems { get; set; }  
}
```

### Properties

| API | Description |
| ------------- | ------------- |
| LeftItems | Gets or sets the items that can be invoked when the control is swiped from the left side. |
| RightItems | Gets or sets the items that can be invoked when the control is swiped from the right side. |
| TopItems | Gets or sets the items that can be invoked when the control is swiped from the top. |
| BottomItems | Gets or sets the items that can be invoked when the control is swiped from the bottom. |

### Examples

```xml
<SwipeView>
	<SwipeView.LeftItems>
		<SwipeItems Mode="Reveal">
			<SwipeItem Text="Favorite" Command="{Binding Favorite}">
				<SwipeItem.Icon>
					<FontIconSource Glyph="&#xE734;"/>
				</SwipeItem.Icon>
			</SwipeItem>
			<SwipeItem Text="Share" Command="{Binding Share}">
				<SwipeItem.Icon>
					<FontIconSource Glyph="&#xE27D;"/>
				</SwipeItem.Icon>
			</SwipeItem>
		</SwipeItems>
	</SwipeView.LeftItems>

	<SwipeView.RightItems>
		<SwipeItems Mode="Execute">
			<SwipeItem Text="Delete" Command="{Binding Delete}">
				<SwipeItem.Icon>
					<FontIconSource Glyph="&#xE74D;"/>
				</SwipeItem.Icon>
			</SwipeItem>
		</SwipeItems>
	</SwipeView.LeftItems>

	<!-- Swipeable content -->
	<Frame>
		<Label Text="Foo" />
	</Frame>
</SwipeView>
```

In this example, the Frame and its contents can be swiped left or right. Swiping to the right will immediately execute the bound "Delete" command. Swiping to the left will reveal two options: "Share" and "Favorite". Tapping on those items will execute their respective commands. 

### Platform Specific

Using a Platform Specific we can control the transition that will be used when opening the SwipeView.

```
public enum SwipeTransitionMode
{
     Drag = -1,
     Reveal = 0
}
```

Available for Android and iOS.

![SwipeTransitionMode](https://user-images.githubusercontent.com/6755973/67283488-0ecbeb80-f4d4-11e9-8bb1-d47c8b7cf00a.gif)

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android
- UWP

### Screenshots ### 
Basic use
![basic-swipeview](https://user-images.githubusercontent.com/6755973/65325128-4e8a7500-dbae-11e9-80b4-afb7448aacd5.gif)

Working with ListView
![listview-swipe](https://user-images.githubusercontent.com/6755973/65325065-2733a800-dbae-11e9-84e2-a1273fbd1ca7.gif)

Working with CollectionView
![collection-swipe](https://user-images.githubusercontent.com/6755973/65325061-2569e480-dbae-11e9-8575-4f013c76ccb2.gif)

### Testing Procedure ###
Launch Core Gallery SwipeView samples.

### PR Checklist ###
<!-- To be completed by reviewers -->

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

PureWeen commented on Xamarin.Forms.Controls/GalleryPages/SwipeViewGalleries/SwipeBehaviorOnInvokedGallery.cs:
```diff
@@ -0,0 +1,81 @@
+	[Preserve(AllMembers = true)]
+	public class SwipeBehaviorOnInvokedGallery : ContentPage
+	{
+		public SwipeBehaviorOnInvokedGallery()
```
If I'm understanding this gallery correctly it seems like the *Auto* isn't working here. It's acting the same as Close

jsuarezruiz replied:
We are using SwipeItems with `Reveal` Mode. In Reveal mode, the SwipeView closes after an item is invoked. So, with `Auto` or `Close` values the SwipeView must close and with `RemainOpen` option must continue open.

PureWeen commented on Xamarin.Forms.Controls/GalleryPages/SwipeViewGalleries/SwipeThresholdGallery.cs:
```diff
@@ -0,0 +1,79 @@
+	[Preserve(AllMembers = true)]
+	public class SwipeThresholdGallery : ContentPage
+	{
+		public SwipeThresholdGallery()
```
Android/ios: The threshold doesn't quite seem right. On iOS if I set the slider to the middle or all the way to the right the action will get invoked if I only swipe to the middle.  

![B93D3AAC-6E34-406D-8BA2-390040C73ADD](https://user-images.githubusercontent.com/5375137/65463678-6c0d4800-de15-11e9-818c-33f468f4d1e9.GIF)


UWP: If I set the threshold all the way to the left then the view still swipes all the way open before executing 


![B176A920-8255-4960-B1C9-9B0A7630593B](https://user-images.githubusercontent.com/5375137/65463602-44b67b00-de15-11e9-905a-b2c78d015d23.GIF)

jsuarezruiz replied:
The SwipeThreshold property is not in the Spec and I have removed it (at least for now).

PureWeen commented on Xamarin.Forms.Controls/GalleryPages/SwipeViewGalleries/SwipeCollectionViewGallery.xaml:
```diff
@@ -0,0 +1,100 @@
+    </ContentPage.Resources>
+    <ContentPage.Content>
+        <Grid>
+            <CollectionView
```
This doesn't look like it's playing well with recycling on Android. If you swipe open an item then scroll down you'll see a bunch of Cells now have an open SwipeView

jsuarezruiz replied:
I have reviewed this case and you are right. I have tested SwipeContainer in UWP and I have applied the same logic. If the scroll start and the SwipeView is open, close it.
![scroll-swipeview](https://user-images.githubusercontent.com/6755973/65589324-23769b80-df89-11e9-80a7-c4de2e95ef94.gif)

jsuarezruiz commented on Xamarin.Forms.Controls/CoreGallery.cs:
```diff
@@ -350,6 +351,8 @@ public override string ToString()
 				new GalleryPageFactory(() => new StepperCoreGalleryPage(), "Stepper Gallery"),
 				new GalleryPageFactory(() => new SwitchCoreGalleryPage(), "Switch Gallery"),
+				new GalleryPageFactory(() => new SwipeViewCoreGalleryPage(), "SwipeView Core Gallery"),
+				new GalleryPageFactory(() => new SwipeViewGallery(), "SwipeView Gallery"),
```
Fixed!.
![fix-swipeview-ios](https://user-images.githubusercontent.com/6755973/65589532-7e0ff780-df89-11e9-92be-e29e13b4f662.gif)

StephaneDelcroix commented on Xamarin.Forms.Core/ContextItem.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace Xamarin.Forms
+{
+	public abstract class ContextItem : ContentView
```
I would reuse MenuItem, and not allow a full ContentView as swipe item. Else we will end up with full pages being rendered for swipe, killing performances.

If a full swipeView with ContentView is required, CarouselView should be used.

PureWeen replied:
Some discussion around this one https://xamarinhq.slack.com/archives/CF98LQ1L3/p1568666466065500 and the reasons for using a view

jsuarezruiz replied:
Ok, now (with the latest changes) we are not using Views. Actually, on all platforms the SwipeItems are created when you start the Swipe and are disposed when you close the SwipeView.
I have reviewed designs (Dribbble, etc.) and I think there are cases where a View fits better. I will update the Spec with some ideas. Maybe we can have a first Release without allowing Views and if it is something demanded we can add support later.

What do you think?. BTW, I have a small prototype adding support to Views and include it would not be complex, if you see it important we can add it in this PR.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,742 @@
+		public static readonly BindableProperty LeftItemsProperty = BindableProperty.Create(nameof(LeftItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.TwoWay, null);
+		public static readonly BindableProperty RightItemsProperty = BindableProperty.Create(nameof(RightItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.TwoWay, null);
+		public static readonly BindableProperty TopItemsProperty = BindableProperty.Create(nameof(TopItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.TwoWay, null);
+		public static readonly BindableProperty BottomItemsProperty = BindableProperty.Create(nameof(BottomItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.TwoWay, null);
```
those collections should be implicitly created, easing the XAML syntax

StephaneDelcroix replied:
this isn't resolved

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,742 @@
+{
+	[ContentProperty("Content")]
+	[RenderWith(typeof(_SwipeViewRenderer))]
+	public class SwipeView : ContentView, ISwipeViewController, IElementConfiguration<SwipeView>
```
either inherit from Layout<View>, or use MenuItem (in that case, ContentView is ok)

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,742 @@
+
+			_gridChild = new List<Element>();
+
+			CompressedLayout.SetIsHeadless(_content, true);
```
COURAGE

jsuarezruiz replied:
:satisfied: This implementation has already changed and is not used.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,742 @@
+		{
+			IsClippedToBounds = true;
+
+			_content = new Grid();
```
doesn't need to pack a grid if you inherit from layout

jsuarezruiz commented:
After reviewing all the feedback, possible improvement points and applying changes (many changes), I think the PR are ready for a new review!.

The main changes have been:
- The logic that was in `SwipeView.cs` (Core) has been moved to renderers on Android and iOS.
- SwipeItems cease to be Views and become native buttons.
- Instead of use simply translations in elements, animations are used now.

Other changes:
- Added a Platform Specific on Android and iOS to indicate the `SwipeTransitionMode`. At the moment I have added two modes: Reveal and Drag.
- Added more samples in Core Gallery.

![SwipeTransitionMode](https://user-images.githubusercontent.com/6755973/67276303-ae817d80-f4c4-11e9-82a6-8b7518c224de.gif)

I have also updated the description of the PR to update the API section.

PureWeen commented:
Marking do not merge based on changes from conversation today

jsuarezruiz commented:
Updated the PR with the following changes:
- Added `CustomSwipeItem`. It is a SwipeItem that allows to use a View with custom content.
- Added another event, `SwipeChanging`.
- New samples in Core Gallery: Using horizontal CollectionView, using CarouselView, using SwipeView events, using CustomSwipeItem, etc.

You can get goodlooking UI designs using CustomSwipeItem!
![favfighters-small](https://user-images.githubusercontent.com/6755973/67697712-45f53d80-f9a9-11e9-81ee-791e69c956b2.gif)

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeDirection.cs:
```diff
@@ -9,4 +9,38 @@ public enum SwipeDirection
 		Down = 8
 	}
+	
+	public static class SwipeDirectionHelper
```
should this needs to be public ?

jsuarezruiz replied:
Added `[EditorBrowsable(EditorBrowsableState.Never)]`

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeItem.cs:
```diff
@@ -0,0 +1,64 @@
+		public event EventHandler<SwipeItemInvokedEventArgs> Invoked;
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void OnInvoked() => Invoked?.Invoke(this, new SwipeItemInvokedEventArgs(this));
```
as the sender is the swipe item, there's no need to pass it again as eventargs

StephaneDelcroix replied:
what does trigger the command ?

jsuarezruiz replied:
Right. I changed the `EventArgs`. The Command is triggered from the Renderers.

StephaneDelcroix replied:
the command and the event should be triggered at the same time. OnInvoked should trigger it

jsuarezruiz replied:
I have applied changes trigger both at the same time.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeItem.cs:
```diff
@@ -0,0 +1,64 @@
+		public event EventHandler<SwipeItemInvokedEventArgs> Invoked;
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void OnInvoked() => Invoked?.Invoke(this, new SwipeItemInvokedEventArgs(this));
```
same

jsuarezruiz replied:
Changed.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeItem.cs:
```diff
@@ -0,0 +1,64 @@
+		public void OnInvoked() => Invoked?.Invoke(this, new SwipeItemInvokedEventArgs(this));
+	}
+
+	public class SwipeItemInvokedEventArgs : EventArgs
```
the sender is the swipe item, no need for an eventargs

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeItem.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace Xamarin.Forms
+{
+	public interface ISwipeItem
```
one public type per file is usually better

jsuarezruiz replied:
Separated in different files.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeItem.cs:
```diff
@@ -0,0 +1,64 @@
+		public void OnInvoked() => Invoked?.Invoke(this, new SwipeItemInvokedEventArgs(this));
+	}
+
+	public class CustomSwipeItem : ContentView, ISwipeItem
```
Not sure we need to expose this, but if we do, that's not a good name

jsuarezruiz replied:
Mnn, do you have any suggestions for the name? I will think about better options.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeItems.cs:
```diff
@@ -0,0 +1,86 @@
+
+namespace Xamarin.Forms
+{
+	public class SwipeItems : BindableObject, IList<ISwipeItem>
```
I don't see the part where you propagate the BindingContext to the children

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,103 @@
+			_platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<SwipeView>>(() => new PlatformConfigurationRegistry<SwipeView>(this));
+		}
+
+		public static readonly BindableProperty LeftItemsProperty = BindableProperty.Create(nameof(LeftItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.TwoWay, null);
```
why are those 2Way ?

jsuarezruiz replied:
Is not necessary, changed.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,103 @@
+			}
+		}
+
+		public class SwipeChangingEventArgs : EventArgs
```
why defining a base class and not using it ?

jsuarezruiz replied:
Is used in the renderers.

StephaneDelcroix replied:
I mean, why don't this inherit from `BaseSwipeEventArgs` ?

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,103 @@
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void SendSwipeEnded(SwipeEndedEventArgs args) => SwipeEnded?.Invoke(this, args);
+
+		public class BaseSwipeEventArgs : EventArgs
```
why isn't this abstract ?

StephaneDelcroix replied:
also, don't do public nested classes

jsuarezruiz replied:
Yes, abstract is perfect in this case.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,103 @@
+			_platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<SwipeView>>(() => new PlatformConfigurationRegistry<SwipeView>(this));
+		}
+
+		public static readonly BindableProperty LeftItemsProperty = BindableProperty.Create(nameof(LeftItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.TwoWay, null);
```
I don't see BindingContext being propagated

jsuarezruiz replied:
Applied changes to add it: https://github.com/xamarin/Xamarin.Forms/blob/4355300b8ded67ae2911782b6a12a2465559917c/Xamarin.Forms.Core/SwipeItems.cs#L108

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeItems.cs:
```diff
@@ -0,0 +1,86 @@
+
+namespace Xamarin.Forms
+{
+	public class SwipeItems : BindableObject, IList<ISwipeItem>
```
this being a BO, and not an Element, prevents parenting. and without parenting, there's no possible tree traversal to resolve styles and resources

jsuarezruiz replied:
Good point!.

jsuarezruiz commented on Xamarin.Forms.Core/ContextItem.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Xamarin.Forms
+{
+	public abstract class ContextItem : MenuItem
```
It is not really necessary, I have removed it for now.

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,103 @@
+		public static readonly BindableProperty LeftItemsProperty = BindableProperty.Create(nameof(LeftItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.OneWay, null);
+		public static readonly BindableProperty RightItemsProperty = BindableProperty.Create(nameof(RightItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.OneWay, null);
+		public static readonly BindableProperty TopItemsProperty = BindableProperty.Create(nameof(TopItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.OneWay, null);
+		public static readonly BindableProperty BottomItemsProperty = BindableProperty.Create(nameof(BottomItems), typeof(SwipeItems), typeof(SwipeView), null, BindingMode.OneWay, null);
```
still no implicit collection support. add a defaultvaluecreator

jsuarezruiz replied:
Added defaultvaluecreator and an extra sample using SwipeItems from resource: https://github.com/xamarin/Xamarin.Forms/blob/swipeview-II/Xamarin.Forms.Controls/GalleryPages/SwipeView/ResourceSwipeItemsGallery.xaml#L49

StephaneDelcroix commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -95,6 +95,13 @@ public SwipeEndedEventArgs(SwipeDirection swipeDirection) : base(swipeDirection)
+		{
+			return ((SwipeView)bindable).SwipeItemsDefaultValueCreator();
+		}
+
```
you can probably inline all that as a single lambda, and copy it in each property

jsuarezruiz commented:
Related PR adding support on Tizen: https://github.com/xamarin/Xamarin.Forms/pull/8542

jsuarezruiz commented:
I have applied changes based on the latest feedback:
- I have renamed `CustomSwipeItem` to `SwipeItemView`.
- Aligned the behavior when clicking on Android and iOS.
- Fixed an Android issue with SwipeItem not rendering occasionally.
- Fixed an Android issue with SwipeItem flickering occasionally.
- The behavior in Android when exceeding the number of visible elements, is now the same as in iOS.
- Implemented Url IconSource. Now in UWP the SwipeItem icon can be an URL.
- Fixed issue on UWP dynamically modifying the SwipeItems (add/remove).

Other points:
- The Layout you have seen in UWP is the expected one.
- In UWP we use the WinUI `SwipeControl` control that only allows one SwipeControl open at the same time. Should we apply the same behavior on Android and iOS?.
- `SwipeItemView` is not currently implemented in UWP (this is probably the most important pending point). Therefore, you cannot do Swipe from the CustomSwipeItem example.

samhouts commented:
@jsuarezruiz  Also, please retarget to 4.4.0

PureWeen commented on Xamarin.Forms.Platform.Android/Renderers/SwipeViewRenderer.cs:
```diff
@@ -0,0 +1,1043 @@
+		{
+			if (e.NewElement != null)
+			{
+				e.NewElement.CloseRequested += OnCloseRequested;
```
Unsubscribe from on OldElement when (e.OldElement != null)

PureWeen commented on Xamarin.Forms.Platform.UAP/SwipeViewRenderer.cs:
```diff
@@ -0,0 +1,484 @@
+				{
+					SetNativeControl(new WSwipeControl());
+
+					Element.LeftItems.CollectionChanged += OnSwipeItemsChanged;
```
Subscribe to these outside the Control check. The Element on this might get swapped out but the Control will stay the same. Unsubscribe from the OldElement if OldElement is not null

PureWeen commented on Xamarin.Forms.Platform.iOS/Renderers/SwipeViewRenderer.cs:
```diff
@@ -0,0 +1,1032 @@
+		{
+			if (e.NewElement != null)
+			{
+				e.NewElement.CloseRequested += OnCloseRequested;
```
if e.OldElement is not null unsubscribe to this event on that element

jsuarezruiz commented on Xamarin.Forms.Core/SwipeView.cs:
```diff
@@ -0,0 +1,135 @@
+			}
+		}
+
+		public class SwipeChangingEventArgs : EventArgs
```
Right. Changed!.

jsuarezruiz commented:
@PureWeen I think I've covered all the feedback:

- Moved `SwipeItemView` to a new class.
- Moved `SwipeEventArgs` to a new class.
- Unsubscribe CollectionChanged from the OldElement.
- Now you can open the SwipeView even if the width of the SwipeItems exceeds the SwipeView width.
- Changes in touch events in Android to have the same behavior as in iOS.

samhouts commented on Xamarin.Forms.Controls/CoreGalleryPages/SwipeViewCoreGalleryPage.cs:
```diff
@@ -0,0 +1,118 @@
+
+		internal SwipeItems GetRevealSwipeItems()
+		{
+			var addSwipteItem = new SwipeItem { BackgroundColor = Color.Green, Text = "Add", IconImageSource = "coffee.png" };
```
```suggestion
			var addSwipeItem = new SwipeItem { BackgroundColor = Color.Green, Text = "Add", IconImageSource = "coffee.png" };
```

samhouts commented on Xamarin.Forms.Controls/CoreGalleryPages/SwipeViewCoreGalleryPage.cs:
```diff
@@ -0,0 +1,118 @@
+				DisplayAlert("SwipeView", "Add Invoked", "OK");
+			};
+
+			var modifySwipteItem = new SwipeItem { BackgroundColor = Color.Orange, Text = "Modify", IconImageSource = "calculator.png" };
```
```suggestion
			var modifySwipeItem = new SwipeItem { BackgroundColor = Color.Orange, Text = "Modify", IconImageSource = "calculator.png" };
```

samhouts commented on Xamarin.Forms.Controls/CoreGalleryPages/SwipeViewCoreGalleryPage.cs:
```diff
@@ -0,0 +1,118 @@
+
+			var modifySwipteItem = new SwipeItem { BackgroundColor = Color.Orange, Text = "Modify", IconImageSource = "calculator.png" };
+
+			modifySwipteItem.Invoked += (sender, e) =>
```
```suggestion
			modifySwipeItem.Invoked += (sender, e) =>
```

samhouts commented on Xamarin.Forms.Controls/CoreGalleryPages/SwipeViewCoreGalleryPage.cs:
```diff
@@ -0,0 +1,118 @@
+
+			var swipeItems = new SwipeItems
+			{
+				addSwipteItem,
```
```suggestion
				addSwipeItem,
```

samhouts commented on Xamarin.Forms.Controls/CoreGalleryPages/SwipeViewCoreGalleryPage.cs:
```diff
@@ -0,0 +1,118 @@
+			var swipeItems = new SwipeItems
+			{
+				addSwipteItem,
+				modifySwipteItem
```
```suggestion
				modifySwipeItem
```

samhouts commented on Xamarin.Forms.Core/PlatformConfiguration/AndroidSpecific/SwipeView.cs:
```diff
@@ -0,0 +1,36 @@
+
+	public enum SwipeTransitionMode
+	{
+		Drag = -1,
```
What's the rationale for starting at -1?

jsuarezruiz replied:
It is not necessary, changed.

samhouts commented on Xamarin.Forms.Core/PlatformConfiguration/iOSSpecific/SwipeView.cs:
```diff
@@ -0,0 +1,36 @@
+
+	public enum SwipeTransitionMode
+	{
+		Drag = -1,
```
What's the rationale for starting at -1?

samhouts commented on Xamarin.Forms.Platform.iOS/Renderers/PhoneMasterDetailRenderer.cs:
```diff
@@ -391,7 +391,11 @@ void UpdatePanGesture()
 			}
 
-			UITouchEventArgs shouldReceive = (g, t) => !(t.View is UISlider);
+   			bool shouldReceive(UIGestureRecognizer g, UITouch t)
```
Is this not an issue on iPad?

jsuarezruiz replied:
We avoid to received the touch in the gesture (MasterDetailPage) if the View is and SwipeView  (or and Slider) on Phone, but is not necessary on the Tablet renderer (because is open).

samhouts commented on Xamarin.Forms.Platform.Android/Renderers/SwipeViewRenderer.cs:
```diff
@@ -0,0 +1,1079 @@
+
+namespace Xamarin.Forms.Platform.Android
+{
+	public class SwipeViewRenderer : ViewRenderer<SwipeView, AView>, GestureDetector.IOnGestureListener
```
I think we should be using the FastRenderers pattern here instead.

samhouts replied:
Please add a comment explaining why we're not using FastRenderers here. :)
